File: cpp/source/Pilot.cpp
#include "Pilot.hpp"

#include "state/pilot/Disconnected.hpp"

#include "backend/ClientBackend.hpp"

#include "core/CoreClient.hpp"

using namespace fm;
using namespace fm::state;

using namespace com::fleetmgr::interfaces;

Pilot::Pilot(Listener& listener, core::https::IHttpsClient& coreClient, const std::string& _certPath) :
    IClient(listener, coreClient, _certPath)
{
    setState(std::make_unique<pilot::Disconnected>(*this, listener, *backend.get()));
}

std::string Pilot::toString() const
{
    return "Pilot";
}

ListDevicesResponse Pilot::listDevices()
{
    return backend->getCore().listDevices();
}

ListDevicesResponse Pilot::listConnectedDevices()
{
    ListDevicesResponse response = listDevices();
    ListDevicesResponse result;
    for (int i = 0; i < response.devices_size(); ++i)
    {
        const Device& d = response.devices(i);
        if (d.connectionstate() == CONNECTED)
        {
            Device* e = result.mutable_devices()->Add();
            e->CopyFrom(d);
        }
    }
    return result;
}

File: java/sdk/Pilot.java
package com.fleetmgr.java.sdk;

import com.fleetmgr.java.sdk.state.pilot.Disconnected;
import com.fleetmgr.interfaces.ConnectionState;
import com.fleetmgr.interfaces.Device;
import com.fleetmgr.interfaces.ListDevicesResponse;

import java.io.IOException;
import java.util.concurrent.ExecutorService;

/**
 * Created by: Bartosz Nawrot
 * Date: 04.09.2018
 * Description:
 */
public class Pilot extends Client {

    public Pilot(String coreAddress, String key, Listener listener,
                 ExecutorService executor) {
        super(coreAddress, key, listener, executor);
        setState(new Disconnected(key, this, coreClient, listener, executor));
    }

    public ListDevicesResponse listDevices() throws IOException {
        return coreClient.listDevices();
    }

    public ListDevicesResponse listConnectedDevices() throws IOException {
        ListDevicesResponse.Builder builder = ListDevicesResponse.newBuilder();
        ListDevicesResponse response = listDevices();
        for (Device d : response.getDevicesList()) {
            if (d.getConnectionState() == ConnectionState.CONNECTED) {
                builder.addDevices(d);
            }
        }
        return builder.build();
    }
}
